### PR TITLE
Integrate f-divergence to DPO

### PR DESCRIPTION
**Pull request by 1485840691**
Related issue: https://github.com/huggingface/trl/issues/1259

1) reverse-kl (current default)
command: examples/scripts/dpo.py --output_dir=dpo_anthropic_hh --model_name_or_path=gpt2 --per_device_train_batch_size 4 --max_steps 1000 --learning_rate 1e-5 --gradient_accumulation_steps 1 --logging_steps 10 --eval_steps 500 --output_dir=dpo_anthropic_hh --warmup_steps 150 --report_to wandb --logging_first_step --no_remove_unused_columns
image
image

2) alpha-divergence w/ alpha=0.5
command: examples/scripts/dpo.py --output_dir=dpo_anthropic_hh --model_name_or_path=gpt2 --per_device_train_batch_size 4 --max_steps 1000 --learning_rate 1e-5 --gradient_accumulation_steps 1 --logging_steps 10 --eval_steps 500 --output_dir=dpo_anthropic_hh --warmup_steps 150 --report_to wandb --logging_first_step --no_remove_unused_columns --f_divergence_type alpha_divergence --f_alpha_divergence_coef 0.5

https://wandb.ai/open_source/huggingface/runs/b943bky2?workspace=user-1485840691
image
image
image
image
image

3) js_divergence
command: examples/scripts/dpo.py --output_dir=dpo_anthropic_hh --model_name_or_path=gpt2 --per_device_train_batch_size 4 --max_steps 1000 --learning_rate 1e-5 --gradient_accumulation_steps 1 --logging_steps 10 --eval_steps 500 --output_dir=dpo_anthropic_hh --warmup_steps 150 --report_to wandb --logging_first_step --no_remove_unused_columns --f_divergence_type js_divergence

image
image

### CODE REVIEW

**Comment by kmn1024** [comment]
Does it make sense to explore a similar change to KTO loss, to allow trading off alignment for diversity there?

**Comment by suanflower** [comment]
Thank you very much for your work, but when using forward KL, the loss becomes 0.

image
